invalid memory address or nil pointer dereference: server/mqtt.go:3074

## Defect

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [X] Included `nats-server -DV` output
 - [ ] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)
 Stack trace should be enough

#### Versions of `nats-server` and affected client libraries used:

nats-server v2.2.6

#### OS/Container environment:

nats:latest

#### Steps or code to reproduce the issue:

Not sure how to make that exhaustive, but main points of the config are

- Operator mode with full nats resolver
- Leaf node connected (the issue is not on the leafnode but on the main cluster server)
- Websocket enabled
- MQTT enabled
- JetStream enabled

```
port: 4222

monitor_port: 8222

jetstream {
  store_dir=nats
}

leafnodes {
  port: 7422
}

websocket {
  port: 4443
  no_tls: true
}

mqtt {
  port: 1883
}

operator: "/nats/config/operator.jwt"
resolver: {
  type: full
  dir: "/nats/accounts"
  allow_delete: false
}
```

#### Expected result:

Normal operation

#### Actual result:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x8a8e23]

goroutine 19443 [running]:
github.com/nats-io/nats-server/server.mqttDeliverMsgCbQos0(0xc000157c80, 0xc0002c4600, 0xc000141000, 0x35, 0x0, 0x0, 0xc0002c4859, 0xef, 0xfc7)
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/mqtt.go:3074 +0x103
github.com/nats-io/nats-server/server.(*client).deliverMsg(0xc0002c4600, 0xc000157c80, 0xc0002c4820, 0x35, 0x1000, 0x0, 0x0, 0x0, 0xc0002c5a81, 0x52, ...)
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:3072 +0xa6f
github.com/nats-io/nats-server/server.(*client).processMsgResults(0xc0002c4600, 0xc0002c8b40, 0xc0006be780, 0xc0002c4859, 0xf1, 0xfc7, 0x0, 0x0, 0x0, 0xc0002c4820, ...)
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:4043 +0x5a5
github.com/nats-io/nats-server/server.(*client).processInboundLeafMsg(0xc0002c4600, 0xc0002c4859, 0xf1, 0xfc7)
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/leafnode.go:2095 +0x278
github.com/nats-io/nats-server/server.(*client).processInboundMsg(0xc0002c4600, 0xc0002c4859, 0xf1, 0xfc7)
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:3392 +0xcd
github.com/nats-io/nats-server/server.(*client).parse(0xc0002c4600, 0xc000663a00, 0x106, 0x200, 0x106, 0x0)
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/parser.go:477 +0x2525
github.com/nats-io/nats-server/server.(*client).readLoop(0xc0002c4600, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/client.go:1160 +0x625
github.com/nats-io/nats-server/server.(*Server).createLeafNode.func1()
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/leafnode.go:921 +0x45
created by github.com/nats-io/nats-server/server.(*Server).startGoRoutine
        /home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:2804 +0xc5
```

I guess this stack trace should be enough to find the issue. If not, let me know and I'll add details about the configuration.

--- a/server/client.go
+++ b/server/client.go
@@ -4035,12 +4035,12 @@ func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, deliver,
 
 		// Remap to the original subject if internal.
 		if sub.icb != nil && sub.rsi {
-			subj = subject
+			dsubj = subject
 		}
 
 		// Normal delivery
 		mh := c.msgHeader(dsubj, creply, sub)
-		didDeliver = c.deliverMsg(sub, subj, creply, mh, msg, rplyHasGWPrefix) || didDeliver
+		didDeliver = c.deliverMsg(sub, dsubj, creply, mh, msg, rplyHasGWPrefix) || didDeliver
 	}
 
 	// Set these up to optionally filter based on the queue lists.

--- a/server/mqtt.go
+++ b/server/mqtt.go
@@ -1698,12 +1698,18 @@ func (as *mqttAccountSessionManager) processSubs(sess *mqttSession, c *client,
 	// Helper that sets the sub's mqtt fields and possibly serialize retained messages.
 	// Assumes account manager and session lock held.
 	setupSub := func(sub *subscription, qos byte) {
-		if sub.mqtt == nil {
-			sub.mqtt = &mqttSub{}
+		subs := []*subscription{sub}
+		if len(sub.shadow) > 0 {
+			subs = append(subs, sub.shadow...)
 		}
-		sub.mqtt.qos = qos
-		if fromSubProto {
-			as.serializeRetainedMsgsForSub(sess, c, sub, trace)
+		for _, sub := range subs {
+			if sub.mqtt == nil {
+				sub.mqtt = &mqttSub{}
+			}
+			sub.mqtt.qos = qos
+			if fromSubProto {
+				as.serializeRetainedMsgsForSub(sess, c, sub, trace)
+			}
 		}
 	}
 

--- a/server/mqtt_test.go
+++ b/server/mqtt_test.go
@@ -683,6 +683,7 @@ func testMQTTWrite(c net.Conn, buf []byte) (int, error) {
 }
 
 func testMQTTConnect(t testing.TB, ci *mqttConnInfo, host string, port int) (net.Conn, *mqttReader) {
+	t.Helper()
 	return testMQTTConnectRetry(t, ci, host, port, 0)
 }
 
@@ -3191,6 +3192,56 @@ func TestMQTTLeafnodeWithoutJSToClusterWithJS(t *testing.T) {
 	// Connect from one server in the cluster check it works from there too.
 	connectAndPublish(o1)
 	testMQTTCheckPubMsg(t, mc2, rc2, "foo", mqttPubQos1, []byte("msg"))
+}
+
+func TestMQTTImportExport(t *testing.T) {
+	conf := createConfFile(t, []byte(`
+		listen: "127.0.0.1:-1"
+		server_name: "mqtt"
+		jetstream {
+			store_dir=org_dir
+		}
+		accounts {
+			org {
+				jetstream: enabled
+				users: [{user: org, password: pwd}]
+				imports = [{stream: {account: "device", subject: "foo"}, prefix: "org"}]
+			}
+			device {
+				users: [{user: device, password: pwd}]
+				exports = [{stream: "foo"}]
+			}
+		}
+		mqtt {
+			listen: "127.0.0.1:-1"
+		}
+		no_auth_user: device
+	`))
+	defer os.Remove(conf)
+	defer os.RemoveAll("org_dir")
+
+	s, o := RunServerWithConfig(conf)
+	defer s.Shutdown()
+
+	mc1, rc1 := testMQTTConnect(t, &mqttConnInfo{clientID: "sub1", user: "org", pass: "pwd", cleanSess: true}, o.MQTT.Host, o.MQTT.Port)
+	defer mc1.Close()
+	testMQTTCheckConnAck(t, rc1, mqttConnAckRCConnectionAccepted, false)
+	testMQTTSub(t, 1, mc1, rc1, []*mqttFilter{{filter: "org/foo", qos: 0}}, []byte{0})
+	testMQTTFlush(t, mc1, nil, rc1)
+
+	mc2, rc2 := testMQTTConnect(t, &mqttConnInfo{clientID: "sub2", user: "org", pass: "pwd", cleanSess: true}, o.MQTT.Host, o.MQTT.Port)
+	defer mc2.Close()
+	testMQTTCheckConnAck(t, rc2, mqttConnAckRCConnectionAccepted, false)
+	testMQTTSub(t, 1, mc2, rc2, []*mqttFilter{{filter: "org/foo", qos: 1}}, []byte{1})
+	testMQTTFlush(t, mc2, nil, rc2)
+
+	nc := natsConnect(t, s.ClientURL())
+	defer nc.Close()
+	natsPub(t, nc, "foo", []byte("msg"))
+
+	// Verify message is received on receiver side.
+	testMQTTCheckPubMsg(t, mc1, rc1, "org/foo", 0, []byte("msg"))
+	testMQTTCheckPubMsg(t, mc2, rc2, "org/foo", 0, []byte("msg"))
 }
 
 func TestMQTTSessionMovingDomains(t *testing.T) {

